### DISCUSSION TITLE
Issue with synchronous children tasks missing Request Context

### DISCUSSION THREAD

**Discussion by nikatlas** [celery/celery#9200] [Q&A]
## Problem 

I have this issue where the request context of the task is dropped when the task calls subsequent tasks in synchronous mode(apply) or with EAGER mode.

## Example
A task like this for example loses its options after the first recursion. 
```python
    @app.task(bind=True)
    def recursive_task(self, *args, retries=2, **kwargs):
        assert self.request.headers.get('test') == 123
        if retries > 0:
            return recursive_task.apply(
                kwargs={"retries": retries - 1},
                headers={'test': 123},
            )
        return True
```

## Expectated behaviour 

My expectation would be the Context to persist. 
I expect to have task information in a bound task regardless of whether it was called synchronously or asynchronously.

Note: 
 I am not sure, if I am missing something, whether a worker should never run in EAGER mode or this is expected somehow. My 
 use-case includes adding custom headers to all celery tasks and allowing propagation of those via a Custom Task.

### tl;dr: Context is dropped.

 
Debugging the celery worker, the task initially had the proper context. 
But on the upcoming recursion  the self.request is empty apart from args, kwargs which are persisted.

This does not happen:
- In my pytest unit tests with TASK_ALWAYS_EAGER flag
- When I run the celery worker in normal/async mode and I use apply_async

It happens only:
- With a celery worker set in EAGER mode.
- With a normal/async worker and using `apply`.

Debugging the code it seems the context is properly persisted up to this point:
https://github.com/celery/celery/blob/78ac69cfed7c485ba32726deaeaf6401d5e7bc1f/celery/app/trace.py#L408-L409

But then at this line where the actual method is being run:

https://github.com/celery/celery/blob/78ac69cfed7c485ba32726deaeaf6401d5e7bc1f/celery/app/trace.py#L453

The stack_protection kicks-in (which seems to be only available in workers:

https://github.com/celery/celery/blob/78ac69cfed7c485ba32726deaeaf6401d5e7bc1f/celery/app/trace.py#L713-L739

This causes a push in the request stack with a poor Context (only args, kwargs)

https://github.com/celery/celery/blob/78ac69cfed7c485ba32726deaeaf6401d5e7bc1f/celery/app/task.py#L407-L414

https://github.com/celery/celery/blob/78ac69cfed7c485ba32726deaeaf6401d5e7bc1f/celery/app/task.py#L1117

Any ideas? or tips? 
Is this expected behavior?

**Comment by nikatlas** [comment]
I made an issue as well with a pytest-celery reproducible test case 😅 

https://github.com/celery/celery/issues/9199
